feat(calendar): introduce before-cell-render event (#926)
Refinitiv/refinitiv-ui: documents/src/pages/elements/calendar.md

const holidays = ['2023-04-07', '2023-04-10', '2023-05-01', '2023-05-18', '2023-05-29'];

calendar.filter = date => !holidays.includes(date);
```
```html
<div style="display:flex">
  <ef-calendar fill-cells view="2023-04" lang="de" id="calendar">
    <div class="holiday" slot="2023-04-07">7</div>
    <div class="holiday" slot="2023-04-10">10</div>
    <div class="holiday" slot="2023-05-01">1</div>
    <div class="holiday" slot="2023-05-18">18</div>
    <div class="holiday" slot="2023-05-29">29</div>
  </ef-calendar>
  <div>
    <h5>Germany Public Holidays 2023</h5>
    <h6>April</h6>
    <p>
      07: Good Friday<br>
      10: Easter Monday
    </p>
    <h6>May</h6>
    <p>
      01: Labour Day<br>
      18: Ascension Day<br>
      29: White Monday
    </p>
  </div>
</div>
```
```css
ef-calendar {
  margin-right: 20px;
  --holiday-background-color: #F5475B;
  --holiday-cell-color: #fff;
}

h5, h6 {
  margin-top: 0px;
}

ef-calendar .holiday {
  background-color: var(--holiday-background-color);
  color: var(--holiday-cell-color);
  width: 100%;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
}
```
::

> For more advanced use cases, you can use `before-cell-render` event to style your slotted cell. The event is fired before each cell is rendered. [Calendar cell model](https://github.com/Refinitiv/refinitiv-ui/blob/v7/packages/elements/src/calendar/types.ts) is provided with the event.

Reviewer: How about a bit shorter version
```suggestion
For more advanced use cases, you can use `before-cell-render` event to get [CalendarCell](https://github.com/Refinitiv/refinitiv-ui/blob/v7/packages/elements/src/calendar/types.ts) to style your slotted cell. The event is fired before and every time each cell is rendered. 
```

  <div class="custom-cell" slot="2023-05-16"></div>
  <div class="custom-cell" slot="2023-05-25"></div>
  <div class="custom-cell" slot="2023-05-31"></div>
  <div class="custom-cell" slot="2023-04"></div>
  <div class="custom-cell" slot="2023-05"></div>
  <div class="custom-cell" slot="2023"></div>
</ef-calendar>
```

```javascript
const calendar = document.getElementById('calendar');

calendar.addEventListener('before-cell-render', (event) => {
  const sourceCalendar = event.target;
  const { cell } = event.detail;
  const customCell = sourceCalendar.querySelector(`[slot="${cell.value}"]`);

  // skip style overriding if there is no content for the cell
  if (!customCell) { return; }

  // use text from component as calendar has built-in locale support
  // for instance, Mai instead of May in German
  customCell.textContent = cell.text;

  // modify classes that match to current cell state
  const keys = ['range', 'selected'];
  const customCellClass = customCell.classList;
  for (const key of keys) {
    cell[key] ? customCellClass.add(key) : customCellClass.remove(key);
  }
});
```

```typescript
import { BeforeCellRenderEvent } from '@refinitiv-ui/elements/calendar';

> const calendar = document.getElementById('calendar');

Reviewer: If we use `querySelector('ef-calendar')` stead of `getElementById` we would have,

```ts
const calendar: Calendar | null
```

instead of 

```ts
const calendar: HTMLElement | null
```

Also, let's recheck `typescript` example and make sure that there is no type errors

Author: code example updated.

  <div class="custom-cell" slot="2023-04"></div>
  <div class="custom-cell" slot="2023-05"></div>
  <div class="custom-cell" slot="2023"></div>
</ef-calendar>
```

```javascript
const calendar = document.querySelector('ef-calendar');

calendar?.addEventListener('before-cell-render', (event) => {
  const sourceCalendar = event.target;
  const { cell } = event.detail;
  const customCell = sourceCalendar.querySelector(`[slot="${cell.value}"]`);

  // skip style overriding if there is no content for the cell
  if (!customCell) { return; }

  // use text from component as calendar has built-in locale support
  // for instance, Mai instead of May in German
  customCell.textContent = cell.text ?? '';

  // modify classes that match to current cell state
  const customCellClass = customCell.classList;
  const keys = ['range', 'selected'];
  for (const key of keys) {
    cell[key] ? customCellClass.add(key) : customCellClass.remove(key);
  }
});
```

```typescript
import { BeforeCellRenderEvent, Calendar, CalendarCell } from '@refinitiv-ui/elements/calendar';

const calendar = document.querySelector<Calendar>('ef-calendar');

calendar?.addEventListener('before-cell-render', (event) => {

>   const sourceCalendar = event.target as Element;

Reviewer: ```suggestion
  const sourceCalendar = event.target as Calendar;
```

  const { cell } = (event as BeforeCellRenderEvent).detail;

>   const customCell = sourceCalendar.querySelector(`[slot="${cell.value}"]`);

Reviewer: TypeScript complains about undefined `cell.value`
```suggestion
  const customCell = sourceCalendar.querySelector(`[slot="${cell.value || ''}"]`);
```

Author: Nice catch! `cell.value` could be `undefined` for empty cells. Let me handle this.

update: remove optional from all props of `CalendarCell` by providing default value. This should be the most user friendly.
